target_chains/solana/programs/pyth-price-publisher/src/instruction.rs

```diff
@@ -0,0 +1,133 @@
+    initialize::initialize, initialize_publisher::initialize_publisher,
+    submit_prices::submit_prices,
+};
+
+/// Seed used to derive the vault account holding funds for initializing
+/// publisher accounts. They cannot afford to pay for allocating accounts
+/// on PythNet without this account helping fund it.
+const VAULT_SEED: &str = "VAULT";
+
+/// Seed used to derive the associated buffer account that publishers can
+/// write their updates into.
+const BUFFER_SEED: &str = "BUFFER";
+
+/// Max number of prices allowed to be stored in a single publisher account.
+const MAX_NUM_PRICES: usize = 128;
+
+#[repr(u8)]
+pub enum Instruction {
+    // key[0] payer     [signer writable]
+    // key[1] vault     [writable]
+    // key[2] system    []
+    Initialize,
+    // key[0] publisher [signer writable]
+    // key[1] buffer    [writable]
+    SubmitPrices,
+    // key[0] publisher []
+    // key[1] vault     [writable]
+    // key[2] buffer    [writable]
+    // key[3] system    []
+    InitializePublisher,
+}
+
+impl Instruction {
+    pub fn parse(input: &[u8]) -> Result<Instruction, ProgramError> {
+        match input.first() {
+            Some(0) => Ok(Instruction::Initialize),
+            Some(1) => Ok(Instruction::SubmitPrices),
+            Some(2) => Ok(Instruction::InitializePublisher),
+            _ => Err(ProgramError::InvalidInstructionData),
+        }
+    }
+}
+
+fn validate_publisher<'a>(
+    account: Option<&AccountInfo<'a>>,
+    require_signer_writable: bool,
+) -> Result<AccountInfo<'a>, ProgramError> {
+    let publisher = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    if require_signer_writable {
+        ensure!(
+            ProgramError::InvalidInstructionData,
+            publisher.is_signer,
+            publisher.is_writable
+        );
+    }
+    Ok(publisher)
+}
+
+fn validate_system<'a>(account: Option<&AccountInfo<'a>>) -> Result<AccountInfo<'a>, ProgramError> {
+    let system = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    ensure!(
+        ProgramError::InvalidInstructionData,
+        system_program::check_id(system.key)
+    );
+    Ok(system)
+}
+
+fn validate_payer<'a>(account: Option<&AccountInfo<'a>>) -> Result<AccountInfo<'a>, ProgramError> {
+    let payer = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    ensure!(
+        ProgramError::InvalidInstructionData,
+        payer.is_signer,
+        payer.is_writable
+    );
+    Ok(payer)
+}
+
+fn pubkey_eq(a: &Pubkey, b: &Pubkey) -> bool {
+    sol_memcmp(a.as_ref(), b.as_ref(), 32) == 0
+}
+
+fn validate_vault<'a>(
+    account: Option<&AccountInfo<'a>>,
+    program_id: &Pubkey,
+) -> Result<(AccountInfo<'a>, u8), ProgramError> {
+    let vault = account
+        .cloned()
+        .ok_or(ProgramError::InvalidInstructionData)?;
+    let vault_pda = Pubkey::find_program_address(&[VAULT_SEED.as_bytes()], program_id);
```

Comment:
find_program_address is inefficient as might try 256 times with different seeds. Since this is constant based on our program id, we can use `declare_id` to define id and have this one as a static one with a test to verify that it matches.

Reply:
Replaced with `create_program_address` and added bumps to the instruction arguments.